File: ooder-agent/src/main/java/net/ooder/agent/skill/scene/llm/ModuleApiRegistry.java
package net.ooder.mvp.skill.scene.llm;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import jakarta.annotation.PostConstruct;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.BiFunction;

@Component
public class ModuleApiRegistry {

    private static final Logger log = LoggerFactory.getLogger(ModuleApiRegistry.class);

    private final Map<String, ModuleApis> modules = new ConcurrentHashMap<>();
    private String currentModule;

    @PostConstruct
    public void init() {
        registerDiscoveryApis();
        registerInstallApis();
        registerActivationApis();
        log.info("[ModuleApiRegistry] Initialized {} modules", modules.size());
    }

    private void registerDiscoveryApis() {
        ModuleApis discovery = new ModuleApis("discovery", "能力发现模块");

        discovery.api("selectCapability")
            .description("选择一个能力")
            .param("capabilityId", "string", "能力ID", true)
            .param("capabilityName", "string", "能力名称", false);

        discovery.api("startScan")
            .description("开始扫描能力")
            .param("method", "string", "发现方式(LOCAL_FS/GITHUB/GITEE/AUTO)", true)
            .param("forceRefresh", "boolean", "是否强制刷新", false);

        discovery.api("filterCapabilities")
            .description("筛选能力")
            .param("type", "string", "能力类型(SCENE/SKILL/COMMUNICATION/AI)", false)
            .param("keyword", "string", "关键词", false)
            .param("installed", "boolean", "是否已安装", false);

        discovery.api("getCapabilityDetail")
            .description("获取能力详情")
            .param("capabilityId", "string", "能力ID", true);

        discovery.api("refreshScan")
            .description("刷新扫描结果")
            .param("method", "string", "发现方式", false);

        discovery.api("getDriverConditions")
            .description("获取驱动条件")
            .param("capabilityId", "string", "能力ID", true);

        modules.put("discovery", discovery);
        log.debug("[ModuleApiRegistry] Registered discovery module with {} APIs", discovery.getApiNames().size());
    }

    private void registerInstallApis() {
        ModuleApis install = new ModuleApis("install", "安装向导模块");

        install.api("startInstall")
            .description("开始安装")
            .param("capabilityId", "string", "能力ID", true)
            .param("capabilityName", "string", "能力名称", false)
            .param("config", "object", "安装配置", false);

        install.api("setConfig")
            .description("设置配置项")
            .param("key", "string", "配置键", true)
            .param("value", "any", "配置值", true);

        install.api("nextStep")
            .description("下一步");

        install.api("prevStep")
            .description("上一步");

        install.api("confirm")
            .description("确认安装");

        install.api("cancel")
            .description("取消安装");

        install.api("checkDependencies")
            .description("检查依赖")
            .param("capabilityId", "string", "能力ID", true);

        modules.put("install", install);
        log.debug("[ModuleApiRegistry] Registered install module with {} APIs", install.getApiNames().size());
    }

    private void registerActivationApis() {
        ModuleApis activation = new ModuleApis("activation", "激活流程模块");

        activation.api("executeStep")
            .description("执行激活步骤")
            .param("stepId", "string", "步骤ID", true)
            .param("data", "object", "步骤数据", false);

        activation.api("skipStep")
            .description("跳过步骤")
            .param("stepId", "string", "步骤ID", true);

        activation.api("selectDriverCondition")
            .description("选择驱动条件")
            .param("conditionId", "string", "条件ID", true);

        activation.api("setParticipants")
            .description("设置参与者")
            .param("participants", "array", "参与者列表", true);

        activation.api("configurePrivateCapabilities")
            .description("配置私有能力")
            .param("enabledCapabilityIds", "array", "启用的能力ID列表", true);

        activation.api("getKey")
            .description("获取安全密钥");

        activation.api("complete")
            .description("完成激活");

        activation.api("cancel")
            .description("取消激活");

        modules.put("activation", activation);
        log.debug("[ModuleApiRegistry] Registered activation module with {} APIs", activation.getApiNames().size());
    }

    public void setCurrentModule(String module) {
        if (modules.containsKey(module)) {
            this.currentModule = module;
            log.info("[ModuleApiRegistry] Current module set to: {}", module);
        } else {
            log.warn("[ModuleApiRegistry] Unknown module: {}", module);
        }
    }

    public String getCurrentModule() {
        return currentModule;
    }

    public Set<String> getAvailableApis(String module) {
        ModuleApis moduleApis = modules.get(module);
        return moduleApis != null ? moduleApis.getApiNames() : Collections.emptySet();
    }

    public Set<String> getCurrentAvailableApis() {
        return getAvailableApis(currentModule);
    }

    public ApiDefinition getApiDefinition(String module, String apiName) {
        ModuleApis moduleApis = modules.get(module);
        return moduleApis != null ? moduleApis.getApi(apiName) : null;
    }

    public Map<String, Object> getModuleInfo(String module) {
        ModuleApis moduleApis = modules.get(module);
        if (moduleApis == null) {
            return Collections.emptyMap();
        }

        Map<String, Object> info = new HashMap<>();
        info.put("name", moduleApis.getName());
        info.put("description", moduleApis.getDescription());
        info.put("apis", moduleApis.getApiDefinitions());
        return info;
    }

    public List<Map<String, Object>> getAllModulesInfo() {
        List<Map<String, Object>> result = new ArrayList<>();
        for (String moduleName : modules.keySet()) {
            result.add(getModuleInfo(moduleName));
        }
        return result;
    }

    public boolean validateApiCall(String module, String apiName) {
        ModuleApis moduleApis = modules.get(module);
        if (moduleApis == null) {
            return false;
        }
        return moduleApis.hasApi(apiName);
    }

    public boolean validateCurrentApiCall(String apiName) {
        return validateApiCall(currentModule, apiName);
    }

    public static class ModuleApis {
        private final String name;
        private final String description;
        private final Map<String, ApiDefinition> apis = new LinkedHashMap<>();

        public ModuleApis(String name, String description) {
            this.name = name;
            this.description = description;
        }

        public ApiDefinition api(String name) {
            ApiDefinition api = new ApiDefinition(name);
            apis.put(name, api);
            return api;
        }

        public String getName() { return name; }
        public String getDescription() { return description; }
        public Set<String> getApiNames() { return apis.keySet(); }
        public ApiDefinition getApi(String name) { return apis.get(name); }
        public boolean hasApi(String name) { return apis.containsKey(name); }

        public List<Map<String, Object>> getApiDefinitions() {
            List<Map<String, Object>> result = new ArrayList<>();
            for (ApiDefinition api : apis.values()) {
                result.add(api.toMap());
            }
            return result;
        }
    }

    public static class ApiDefinition {
        private final String name;
        private String description;
        private final Map<String, ParamDefinition> parameters = new LinkedHashMap<>();

        public ApiDefinition(String name) {
            this.name = name;
        }

        public ApiDefinition description(String description) {
            this.description = description;
            return this;
        }

        public ApiDefinition param(String name, String type, String description, boolean required) {
            parameters.put(name, new ParamDefinition(name, type, description, required));
            return this;
        }

        public String getName() { return name; }
        public String getDescription() { return description; }
        public Map<String, ParamDefinition> getParameters() { return parameters; }

        public Map<String, Object> toMap() {
            Map<String, Object> map = new HashMap<>();
            map.put("name", name);
            map.put("description", description);

            List<Map<String, Object>> params = new ArrayList<>();
            for (ParamDefinition param : parameters.values()) {
                params.add(param.toMap());
            }
            map.put("parameters", params);

            return map;
        }
    }

    public static class ParamDefinition {
        private final String name;
        private final String type;
        private final String description;
        private final boolean required;

        public ParamDefinition(String name, String type, String description, boolean required) {
            this.name = name;
            this.type = type;
            this.description = description;
            this.required = required;
        }

        public String getName() { return name; }
        public String getType() { return type; }
        public String getDescription() { return description; }
        public boolean isRequired() { return required; }

        public Map<String, Object> toMap() {
            Map<String, Object> map = new HashMap<>();
            map.put("name", name);
            map.put("type", type);
            map.put("description", description);
            map.put("required", required);
            return map;
        }
    }
}

File: ooder-Nexus-Enterprise/skills/_system/skill-llm-chat/src/main/resources/static/console/js/module-api-registry.js
(function(global) {
    'use strict';

    var ModuleApiRegistry = {
        currentModule: null,
        modules: {},
        handlers: {},

        init: function() {
            this.registerDiscoveryApis();
            this.registerInstallApis();
            this.registerActivationApis();
            console.log('[ModuleApiRegistry] Initialized with modules:', Object.keys(this.modules));
        },

        registerDiscoveryApis: function() {
            var discovery = {
                id: 'discovery',
                name: '能力发现模块',
                apis: {}
            };

            discovery.apis.startScan = {
                name: 'startScan',
                description: '开始扫描能力',
                params: {
                    method: { type: 'string', description: '发现方式(LOCAL_FS/GITHUB/GITEE/AUTO)', required: true },
                    forceRefresh: { type: 'boolean', description: '是否强制刷新', required: false }
                },
                execute: function(params) {
                    if (typeof CapabilityDiscovery !== 'undefined') {
                        return CapabilityDiscovery.startScan(params.method, params.forceRefresh);
                    }
                    return { error: 'CapabilityDiscovery not available' };
                }
            };

            discovery.apis.selectCapability = {
                name: 'selectCapability',
                description: '选择一个能力',
                params: {
                    capabilityId: { type: 'string', description: '能力ID', required: true }
                },
                execute: function(params) {
                    if (typeof CapabilityDiscovery !== 'undefined') {
                        return CapabilityDiscovery.selectCapability(params.capabilityId);
                    }
                    return { error: 'CapabilityDiscovery not available' };
                }
            };

            discovery.apis.filterCapabilities = {
                name: 'filterCapabilities',
                description: '筛选能力',
                params: {
                    type: { type: 'string', description: '能力类型(SCENE/SKILL/COMMUNICATION/AI)', required: false },
                    keyword: { type: 'string', description: '关键词', required: false },
                    installed: { type: 'boolean', description: '是否已安装', required: false }
                },
                execute: function(params) {
                    if (typeof CapabilityDiscovery !== 'undefined') {
                        return CapabilityDiscovery.filterCapabilities(params);
                    }
                    return { error: 'CapabilityDiscovery not available' };
                }
            };

            discovery.apis.getCapabilityDetail = {
                name: 'getCapabilityDetail',
                description: '获取能力详情',
                params: {
                    capabilityId: { type: 'string', description: '能力ID', required: true }
                },
                execute: function(params) {
                    if (typeof CapabilityDiscovery !== 'undefined') {
                        return CapabilityDiscovery.getCapabilityDetail(params.capabilityId);
                    }
                    return { error: 'CapabilityDiscovery not available' };
                }
            };

            discovery.apis.getDriverConditions = {
                name: 'getDriverConditions',
                description: '获取驱动条件',
                params: {
                    capabilityId: { type: 'string', description: '能力ID', required: true }
                },
                execute: function(params) {
                    return fetch('/api/v1/discovery/capabilities/detail/' + params.capabilityId + '/driver-conditions')
                        .then(function(res) { return res.json(); });
                }
            };

            this.modules.discovery = discovery;
        },

        registerInstallApis: function() {
            var install = {
                id: 'install',
                name: '安装向导模块',
                apis: {}
            };

            install.apis.startInstall = {
                name: 'startInstall',
                description: '开始安装',
                params: {
                    capabilityId: { type: 'string', description: '能力ID', required: true },
                    config: { type: 'object', description: '安装配置', required: false }
                },
                execute: function(params) {
                    if (typeof InstallWizard !== 'undefined') {
                        return InstallWizard.startInstall(params.capabilityId, params.config);
                    }
                    return { error: 'InstallWizard not available' };
                }
            };

            install.apis.setConfig = {
                name: 'setConfig',
                description: '设置配置项',
                params: {
                    key: { type: 'string', description: '配置键', required: true },
                    value: { type: 'any', description: '配置值', required: true }
                },
                execute: function(params) {
                    if (typeof InstallWizard !== 'undefined') {
                        return InstallWizard.setConfig(params.key, params.value);
                    }
                    return { error: 'InstallWizard not available' };
                }
            };

            install.apis.nextStep = {
                name: 'nextStep',
                description: '下一步',
                params: {},
                execute: function(params) {
                    if (typeof InstallWizard !== 'undefined') {
                        return InstallWizard.nextStep();
                    }
                    return { error: 'InstallWizard not available' };
                }
            };

            install.apis.confirm = {
                name: 'confirm',
                description: '确认安装',
                params: {},
                execute: function(params) {
                    if (typeof InstallWizard !== 'undefined') {
                        return InstallWizard.confirm();
                    }
                    return { error: 'InstallWizard not available' };
                }
            };

            this.modules.install = install;
        },

        registerActivationApis: function() {
            var activation = {
                id: 'activation',
                name: '激活流程模块',
                apis: {}
            };

            activation.apis.executeStep = {
                name: 'executeStep',
                description: '执行激活步骤',
                params: {
                    stepId: { type: 'string', description: '步骤ID', required: true },
                    data: { type: 'object', description: '步骤数据', required: false }
                },
                execute: function(params) {
                    if (typeof ActivationFlow !== 'undefined') {
                        return ActivationFlow.executeStep(params.stepId, params.data);
                    }
                    return { error: 'ActivationFlow not available' };
                }
            };

            activation.apis.skipStep = {
                name: 'skipStep',
                description: '跳过步骤',
                params: {
                    stepId: { type: 'string', description: '步骤ID', required: true }
                },
                execute: function(params) {
                    if (typeof ActivationFlow !== 'undefined') {
                        return ActivationFlow.skipStep(params.stepId);
                    }
                    return { error: 'ActivationFlow not available' };
                }
            };

            activation.apis.selectDriverCondition = {
                name: 'selectDriverCondition',
                description: '选择驱动条件',
                params: {
                    conditionId: { type: 'string', description: '条件ID', required: true }
                },
                execute: function(params) {
                    if (typeof ActivationFlow !== 'undefined') {
                        return ActivationFlow.selectDriverCondition(params.conditionId);
                    }
                    return { error: 'ActivationFlow not available' };
                }
            };

            activation.apis.complete = {
                name: 'complete',
                description: '完成激活',
                params: {},
                execute: function(params) {
                    if (typeof ActivationFlow !== 'undefined') {
                        return ActivationFlow.complete();
                    }
                    return { error: 'ActivationFlow not available' };
                }
            };

            this.modules.activation = activation;
        },

        setCurrentModule: function(moduleId) {
            if (this.modules[moduleId]) {
                this.currentModule = moduleId;
                console.log('[ModuleApiRegistry] Current module set to:', moduleId);
                this.notifyModuleChange(moduleId);
            } else {
                console.warn('[ModuleApiRegistry] Unknown module:', moduleId);
            }
        },

        getCurrentModule: function() {
            return this.currentModule;
        },

        getModule: function(moduleId) {
            return this.modules[moduleId];
        },

        getAvailableApis: function(moduleId) {
            var module = this.modules[moduleId || this.currentModule];
            if (!module) return {};
            return module.apis;
        },

        getApiNames: function(moduleId) {
            var apis = this.getAvailableApis(moduleId);
            return Object.keys(apis);
        },

        getApiDefinition: function(moduleId, apiName) {
            var apis = this.getAvailableApis(moduleId);
            return apis[apiName];
        },

        executeApi: function(apiName, params) {
            if (!this.currentModule) {
                return Promise.reject(new Error('No current module set'));
            }

            var api = this.getApiDefinition(this.currentModule, apiName);
            if (!api) {
                return Promise.reject(new Error('API not found: ' + apiName + ' in module: ' + this.currentModule));
            }

            try {
                var result = api.execute(params || {});
                if (result && typeof result.then === 'function') {
                    return result;
                }
                return Promise.resolve(result);
            } catch (e) {
                return Promise.reject(e);
            }
        },

        validateApiCall: function(apiName, moduleId) {
            if (moduleId !== this.currentModule) {
                return { valid: false, error: 'Module mismatch. Current: ' + this.currentModule + ', Requested: ' + moduleId };
            }

            var api = this.getApiDefinition(moduleId, apiName);
            if (!api) {
                return { valid: false, error: 'API not found: ' + apiName };
            }

            return { valid: true };
        },

        validateParams: function(apiName, params) {
            var api = this.getApiDefinition(this.currentModule, apiName);
            if (!api) return { valid: false, error: 'API not found' };

            var missingParams = [];
            for (var paramName in api.params) {
                if (api.params[paramName].required && (!params || params[paramName] === undefined)) {
                    missingParams.push(paramName);
                }
            }

            if (missingParams.length > 0) {
                return { valid: false, error: 'Missing required params: ' + missingParams.join(', ') };
            }

            return { valid: true };
        },

        getModuleContext: function() {
            var module = this.modules[this.currentModule];
            if (!module) return null;

            return {
                moduleId: module.id,
                moduleName: module.name,
                availableApis: this.getApiNames(),
                apiDefinitions: this.getAvailableApis()
            };
        },

        onModuleChange: function(callback) {
            this.handlers.moduleChange = this.handlers.moduleChange || [];
            this.handlers.moduleChange.push(callback);
        },

        notifyModuleChange: function(moduleId) {
            var handlers = this.handlers.moduleChange || [];
            handlers.forEach(function(callback) {
                try {
                    callback(moduleId);
                } catch (e) {
                    console.error('[ModuleApiRegistry] Handler error:', e);
                }
            });
        },

        getApiSchema: function(moduleId) {
            var module = this.modules[moduleId || this.currentModule];
            if (!module) return null;

            var schema = {
                module: module.id,
                name: module.name,
                apis: []
            };

            for (var apiName in module.apis) {
                var api = module.apis[apiName];
                var apiSchema = {
                    name: api.name,
                    description: api.description,
                    parameters: []
                };

                for (var paramName in api.params) {
                    apiSchema.parameters.push({
                        name: paramName,
                        type: api.params[paramName].type,
                        description: api.params[paramName].description,
                        required: api.params[paramName].required
                    });
                }

                schema.apis.push(apiSchema);
            }

            return schema;
        }
    };

    if (document.readyState === 'loading') {
        document.addEventListener('DOMContentLoaded', function() {
            ModuleApiRegistry.init();
        });
    } else {
        ModuleApiRegistry.init();
    }

    global.ModuleApiRegistry = ModuleApiRegistry;
})(window);
